rfqservice/negotiator.go

```diff
@@ -0,0 +1,279 @@
+package rfqservice
+
+import (
+	"encoding/hex"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/rfqmsg"
+)
+
+// NegotiatorCfg holds the configuration for the negotiator.
+type NegotiatorCfg struct {
+	// PriceOracle is the price oracle that the negotiator will use to
+	// determine whether a quote is accepted or rejected.
+	PriceOracle PriceOracle
+}
+
+// Negotiator is a struct that handles the negotiation of quotes. It is a RFQ
+// subsystem. It determines whether a quote is accepted or rejected.
+type Negotiator struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the negotiator.
+	cfg NegotiatorCfg
+
+	// outgoingMessages is a channel which is populated with outgoing peer
+	// messages.
+	outgoingMessages chan<- rfqmsg.OutgoingMsg
+
+	// assetSellOffers is a map (keyed on asset) that holds the asset sell
+	// offers that the negotiator has received.
+	assetSellOffers map[string]AssetSellOffer
+
+	assetGroupSellOffers map[string]AssetSellOffer
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewNegotiator creates a new quote negotiator.
+func NewNegotiator(cfg NegotiatorCfg,
+	outgoingMessages chan<- rfqmsg.OutgoingMsg) (*Negotiator, error) {
+
+	return &Negotiator{
+		cfg:              cfg,
+		outgoingMessages: outgoingMessages,
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// queryPriceOracle queries the price oracle for the asking price.
+func (n *Negotiator) queryPriceOracle(
+	req rfqmsg.Request) (*OracleAskingPriceResponse, error) {
+
+	ctx, cancel := n.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	res, err := n.cfg.PriceOracle.QueryAskingPrice(
+		ctx, req.AssetID, req.AssetGroupKey, req.AssetAmount,
+		&req.BidPrice,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query price oracle: %w", err)
+	}
+
+	return res, nil
+}
+
+// handlePriceOracleResponse handles a response from the price oracle.
+func (n *Negotiator) handlePriceOracleResponse(request rfqmsg.Request,
+	oracleResponse *OracleAskingPriceResponse) error {
+
+	// If the asking price is nil, then we will return the error message
+	// supplied by the price oracle.
+	if oracleResponse.AskingPrice == nil {
+		rejectMsg := rfqmsg.NewRejectMsg(
+			request.Peer, request.ID, *oracleResponse.Err,
+		)
+		var msg rfqmsg.OutgoingMsg = &rejectMsg
+
+		sendSuccess := fn.SendOrQuit(n.outgoingMessages, msg, n.Quit)
```

Comment:
Ah I see now. Alternatively, we can have the outgoing messages be some generic interface that we can use to send messages. then w/e concrete implementation can use an underlying channel for synchronization purposes.